### PR TITLE
Fix #397 wrong cell voltage reading daly bms

### LINKED ISSUE #397: Wrong cell voltage reading Daly BMS
**Describe the bug**
I have 12S Daly BMS and function read_cells_volts always returning from here

if cells_volts_data is False:
logger.warning("read_cells_volts")
return False

I had to change
cellvolt_buffer = b"\xA5\x40\x94\x08\x00\x00\x00\x00\x00\x00\x00\x00\x82\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
to
cellvolt_buffer = b"\xA5\x40\x94\x08\x00\x00\x00\x00\x00\x00\x00\x00\x82"

becuse BMS doesn't want to answer to that so huge command

Second thing because cell_count is 12  maxFrame were counted wrong 

            if (int(self.cell_count) % 3) == 0:
                maxFrame = (int(self.cell_count / 3))
            else:
                maxFrame = (int(self.cell_count / 3) + 1)

Third thing lenFixed should be 13 not 12 because each returned frame with cell voltage have own CRC  
            lenFixed = (maxFrame * 13) # 0xA5, 0x01, 0x95, 0x08 + 1 byte frame + 6 byte data + 1byte reserved

### PR DESCRIPTION
**Pull request by transistorgit**
this pull requests addresses some bugs found inside daly.py and utils.py, and were inspired by the findings of @ats-rozruch

- the command buffer was much to big and caused problems with RS485 communication
- the buffer length calculation was not working for all cell string legths
- some other wrong numbers in check sum length, field length etc
- checksum check was missing for cell voltage read function
- in utils.py was a wrong calculation of packet lenth that didn't took the different behaviour of usages with and withoud fixed package length

fixes #397

### CODE REVIEW

**Comment by transistorgit** [comment]
tested with 2 daly smart bms 16s. One with RS232-UART, one with RS485  connection.

**Comment by towatai** [comment]
Works without problems with two 150A / 8S connected over RS485.

File: etc/dbus-serialbattery/daly.py [hunk 1]
```diff
@@ -20,7 +20,7 @@ def __init__(self, port, baud, address):
 
     # command bytes [StartFlag=A5][Address=40][Command=94][DataLength=8][8x zero bytes][checksum]
     command_base = b"\xA5\x40\x94\x08\x00\x00\x00\x00\x00\x00\x00\x00\x81"
-    cellvolt_buffer = b"\xA5\x40\x94\x08\x00\x00\x00\x00\x00\x00\x00\x00\x82\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
+    cellvolt_buffer = b"\xA5\x40\x94\x08\x00\x00\x00\x00\x00\x00\x00\x00\x82"
```
Comment (transistorgit) [inline etc/dbus-serialbattery/daly.py:23]: cellvolt buffer makes no sense. the read_serial function returns a new buffer

File: etc/dbus-serialbattery/daly.py [hunk 2]
```diff
@@ -246,13 +246,16 @@ def read_cells_volts(self, ser):
             buffer[1] = self.command_address[0]  # Always serial 40 or 80
             buffer[2] = self.command_cell_volts[0]
 
-            maxFrame = int(self.cell_count / 3) + 1
+            if (int(self.cell_count) % 3) == 0:
+                maxFrame = (int(self.cell_count / 3))
+            else:
+                maxFrame = (int(self.cell_count / 3) + 1)
```
Comment (transistorgit) [inline etc/dbus-serialbattery/daly.py:252]: fix frame calculation for other cell counts

File: etc/dbus-serialbattery/daly.py [hunk 3]
```diff
@@ -246,13 +246,16 @@ def read_cells_volts(self, ser):
...
             lenFixed = (
-                maxFrame * 12
-            )  # 0xA5, 0x01, 0x95, 0x08 + 1 byte frame + 6 byte data + 1byte reserved
+                maxFrame * 13
+            )  # 0xA5, 0x01, 0x95, 0x08 + 1 byte frame + 6 byte data + 1byte reserved + chksum
```
Comment (transistorgit) [inline etc/dbus-serialbattery/daly.py:255]: fix sentence length, as there is a checksum

File: etc/dbus-serialbattery/daly.py [hunk 4]
```diff
@@ -246,13 +246,16 @@ def read_cells_volts(self, ser):
...
 
             cells_volts_data = read_serialport_data(
-                ser, buffer, self.LENGTH_POS, self.LENGTH_CHECK, lenFixed
+                ser, buffer, self.LENGTH_POS, 0, lenFixed
```
Comment (transistorgit) [inline etc/dbus-serialbattery/daly.py:258]: the normal call is not working on these multi-sentence transmissions. so just receive the full length and calculate the checksums in the cellvolts function later

**Comment by transistorgit** [comment]
I reverted the read_serialport_data() function back to the old state. Changes are in daly.py locally now.

**Review by Louisvdw — APPROVED** [review]

**Pull request merged**